### PR TITLE
add fine-tune-llama2 template for AIOA

### PR DESCRIPTION
**Pull request by scottsun94**
- The training logic should be exactly the same as the current AICA template code ([master branch](https://github.com/ray-project/ray/blob/master/doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py))
- refactored the bash script. Change the CLI parameter parsing logic into python code and merge it with the main python file
- Simplified the readme file for users to get started within 2 mins.

### CODE REVIEW

**Comment by scottsun94** [comment]
@ArturNiederfahrenhorst I have refactored the bash script into python and merge it with the main python script. Can you review that part?

File: templates/fine-tune-llama2/train.py [hunk 1]
```diff
@@ -0,0 +1,871 @@
...
+            enumerate(train_dataloader), total=train_ds_len // batch_size + 1
+        ):
+
+            # We could avoid this line since we set the accelerator with
+            # `device_placement=True`.
+            with accelerator.accumulate(model):
+                s_fwd = time.time()
+                outputs = model(**batch)
+                loss = outputs.loss
+                loss_sum += loss
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/train.py:406]: We had a bug fix the other day, where this line became `loss.item()`. Please make sure that the code in this PR is based off of the freshest OSS template commit!

File: templates/fine-tune-llama2/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+python train.py --size=7b --as-test
+```
+
+Model checkpoints will be stored under `{user's first name}/ft_llms_with_deepspeed/` in the cloud stroage bucket created for your Anyscale account. The full path will be printed in the output after the training is completed.
+
+
+## Step 2: Test the fine-tuned model
+
+
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/README.md:20]: Something missing? 🐳

File: configs/fine-tune-llama2/aws.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+head_node_type:
+  name: head_node_type
+  instance_type: m5.xlarge
+
+worker_node_types:
+- name: gpu_worker
+  instance_type: g5.4xlarge
+  min_workers: 0
+  max_workers: 100
```
[the remark this answers is not in the document]
Comment (scottsun94) [reply configs/fine-tune-llama2/aws.yaml:9]: This is an arbitrary number. I just give a high enough number so that users won't run into autoscaling limit issue if they custom stuff or run something not related to this template...

File: templates/fine-tune-llama2/deepspeed_configs/zero_3_llama_2_7b.json [hunk 4]
```diff
@@ -0,0 +1,35 @@
+{
```
[the remark this answers is not in the document]
Comment (scottsun94) [reply templates/fine-tune-llama2/deepspeed_configs/zero_3_llama_2_7b.json:1]: I've made sure that the configs are exactly the same as the master.

File: templates/fine-tune-llama2/merge_lora_weights.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+    print(f"Saved merged model to {args.output_path} after {time.time() - s3} seconds.")
+    print(f"This script took {time.time() - s} seconds to execute.")
+
+    if TEST_EVAL:
+        test_eval(model, tokenizer)
+
+
+if __name__ == "__main__":
+    main()
```
[the remark this answers is not in the document]
Comment (scottsun94) [reply templates/fine-tune-llama2/merge_lora_weights.py]: Code is the same as master

File: templates/fine-tune-llama2/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+# dataset
+# evaluate
+# wandb
+# pytorch-lightning
+# protobuf<3.21.0
+# torchmetrics
+# lm_eval==0.3.0
+# tiktoken==0.1.2
+# sentencepiece
```
[the remark this answers is not in the document]
Comment (scottsun94) [reply templates/fine-tune-llama2/requirements.txt]: Decided to use the same docker image. Removing this entirely

File: templates/fine-tune-llama2/train.py (hunk 1 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/train.py]: Fixed. Training code is now the same as the master

File: templates/fine-tune-llama2/README.md (hunk 2 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/README.md:20]: Yeah.. This is a placeholder. I'm still trying to figure out how to load the checkpoint locally and run it.

File: templates/fine-tune-llama2/create_dataset.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
+from datasets import load_dataset
```
[the remark this answers is not in the document]
Comment (scottsun94) [reply templates/fine-tune-llama2/create_dataset.py:1]: Done

File: templates/fine-tune-llama2/README.md (hunk 2 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/README.md:25]: I removed this step for now. Need your help to write the code for this step. Resolve it for now.

File: templates/fine-tune-llama2/README.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
+# Fine-tuning Llama-2 models with Deepspeed, Accelerate, and Ray Train
+This template shows you how to fine-tune Llama-2 models. 
+
+## Step 1: Fine-tune Llama-2-7B model
+- Open a terminal in VS Code (ctl+shift+`). 
+- Run the command below to kick off fine-tuning with dummy data, Grade School Math 8k (GSM8K) dataset.
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/README.md:6]: We should put a link here. Since this is meant to be available to absolute entry level, folks might not known that this is a generally known dataset and they should look at it to learn what their LLM is learning and able to do.

File: templates/fine-tune-llama2/train.py [hunk 9]
```diff
@@ -0,0 +1,871 @@
...
+)
+
+OPTIM_BETAS = (0.9, 0.999)
+OPTIM_EPS = 1e-8
+NUM_WARMUP_STEPS = 10
+OPTIM_WEIGHT_DECAY = 0.0
+ATTENTION_LAYER_NAME = "self_attn"
+
+
+def get_expected_lora_num_parameters(
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/train.py:55]: @scottsun94 Please remove this function and its usage.

File: templates/fine-tune-llama2/train.py [hunk 10]
```diff
@@ -0,0 +1,871 @@
...
+
+    return sum_params
+
+
+def get_number_of_params(model: nn.Module):
+    sum = 0
+    for name, param in model.named_parameters():
+        if param.requires_grad:
+            sum += param.numel()
+    return sum
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/train.py:115]: We also don't need this anymore.

Comment (ArturNiederfahrenhorst) [reply templates/fine-tune-llama2/train.py:115]: Let's just remove these to declutter this code. We pin peft anyway, so nothing should go wrong here.

File: templates/fine-tune-llama2/train.py [hunk 11]
```diff
@@ -0,0 +1,871 @@
...
+    parser.add_argument(
+        "--output_dir", 
+        type=str, 
+        default="/mnt/local_storage",
+        help="Path to output directory."
+    )
+
+    # parser.add_argument(
+    #     "--model_name", default="meta-llama/Llama-2-7b-chat-hf", type=str
+    # )
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/train.py:647]: I think we can remove this if it's not needed.

File: templates/fine-tune-llama2/train.py [hunk 12]
```diff
@@ -0,0 +1,871 @@
...
+
+
+def main():
+    print("Fine-tuning model...")
+    args = parse_args()
+
+    # Create the config with args for training.
+    config = vars(args)
+
+    SIZE = "7B" # Default model size
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/train.py:718]: 7b is already the default argument for the --size paramter.
Also, It should be 7b instead of 7B.

File: templates/fine-tune-llama2/train.py [hunk 13]
```diff
@@ -0,0 +1,871 @@
...
+    LR = 5e-6 # Default learning rate for full-parameter fine-tuning
+
+    # Process arguments
+    for arg in sys.argv[1:]:
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/train.py:722]: Let's just use args here instead of sys.argv?
argparse nicely encapsulates this functionality.
You can just do `args.size`.

File: templates/fine-tune-llama2/train.py [hunk 14]
```diff
@@ -0,0 +1,871 @@
...
+
+    # Set up runtime_env for Ray Train
+    ray.init(
+        runtime_env={
+            "env_vars": {
+                "HF_HOME": "/mnt/local_storage/.cache/huggingface",
+                "RAY_AIR_LOCAL_CACHE_DIR": os.environ["RAY_AIR_LOCAL_CACHE_DIR"],
+            },
+            # This is not needed to run in Workspace. However, it might be needed for other environments.
+            # "working_dir": ".",
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/train.py:796]: According to our offline discussion, this can go.

File: templates/fine-tune-llama2/train.py [hunk 15]
```diff
@@ -0,0 +1,871 @@
...
+    artifact_storage = os.environ.get("ANYSCALE_ARTIFACT_STORAGE", "artifact_storage")
+    user_name = re.sub(r"\s+", "__", os.environ.get("ANYSCALE_USERNAME", "user"))
+    model_name=config.get("model_name")
+    storage_path = (
+        f"{artifact_storage}/{user_name}/ft_llms_with_deepspeed/{model_name}"
+    )
+
+    # trial_name = f"{model_name}".split("/")[-1]
+    # if config["lora"]:
+    #     trial_name += "-lora"
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/train.py:822]: This can also go, right?

File: templates/fine-tune-llama2/train.py [hunk 16]
```diff
@@ -0,0 +1,871 @@
...
+        print("Best checkpoint is stored at:")
+        print(best_checkpoint)
+        print(f"With perplexity: {best_checkpoint_metrics['perplexity']}")
+
+        print("Process completed.")
+
+    except Exception as e:
+        print("Failed to fine-tune the model. Exiting...")
+        print(f"Error: {e}")
+        exit()  # This will stop the script
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/train.py:866]: We can also use `sys.exit()` here. At least for consistency but it's also the intended pattern.

**Comment by ArturNiederfahrenhorst** [comment]
Awesome! I have some minor nits and requests for small cleanups but the thing as a whole seems in very good shape to me 👍

File: templates/fine-tune-llama2/README.md (hunk 8 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/README.md:6]: which link should I use? This one? https://github.com/openai/grade-school-math

File: templates/fine-tune-llama2/train.py (hunk 9 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/train.py:55]: Done

File: templates/fine-tune-llama2/train.py (hunk 10 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/train.py:115]: Removed this function as well

File: templates/fine-tune-llama2/train.py (hunk 11 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/train.py:647]: Done

File: templates/fine-tune-llama2/train.py (hunk 12 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/train.py:718]: Changed it to an empty variable.

File: templates/fine-tune-llama2/train.py (hunk 13 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/train.py:722]: Done.

File: templates/fine-tune-llama2/train.py (hunk 14 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/train.py:796]: Done

File: templates/fine-tune-llama2/train.py (hunk 15 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/train.py:822]: It seems so. I'm not very sure but:
-  it's not used in other places.
-  it works fine without it

let's remove for now

File: templates/fine-tune-llama2/train.py (hunk 16 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/train.py:866]: Done.

File: configs/fine-tune-llama2/gce.yaml [hunk 17]
```diff
@@ -0,0 +1,10 @@
+head_node_type:
+  name: head_node_type
+  instance_type: n1-standard-4
+
+worker_node_types:
+- name: gpu_worker
+  instance_type: g2-standard-16-nvidia-l4-1
```
Comment (ArturNiederfahrenhorst) [inline configs/fine-tune-llama2/gce.yaml:7]: The specs make sense. Can you still ACC here that you've tested these configurations? Meaning that you've run a full-parameter FT job of 7B, 13B and 70B for the AWS and GCE yamls here, please?

File: templates/fine-tune-llama2/README.md (hunk 8 above)
Comment (ArturNiederfahrenhorst) [reply templates/fine-tune-llama2/README.md:6]: Let's use this one: https://huggingface.co/datasets/gsm8k

File: templates/fine-tune-llama2/README.md [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+python train.py --size=13b --as-test
+```
+
+
+## Step 3: Use your custom data
+- Replace the contents in `./data/train.jsonl` with your own training data
+- (Optional) Replace the contents in `./data/test.jsonl` with your own test data if any.
+- (Optional) Add special token in `./data/tokens.json` if any.
+
+Use the same command to training with your own data.
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/README.md:38]:
```suggestion
Use the same command to train with your own data.
```

File: templates/fine-tune-llama2/train.py [hunk 19]
```diff
@@ -0,0 +1,776 @@
...
+            "num_devices": ND,
+            "num_epochs": args.num_epochs,
+            "output_dir": args.output_dir,
+            "seed": 42,
+            "special_token_path": args.special_token_path,
+            "stop_perplexity": args.stop_perplexity,
+            "train_path": args.train_path,
+            "test_path": args.test_path,
+        }
+    )
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llama2/train.py:684]: Let's follow the common Python convention that only constants are fully in capitalized letters.
The capital letters are an artifact of this coming from a bash script.

Let's run all model sizes once with LoRA enabled please to verify that this works.

```suggestion
    size = args.size # Variable to store the model size
    lr = 5e-6 # Default learning rate for full-parameter fine-tuning

    # Adjust learning rate if LoRA is enabled
    if args.lora:
        lr = 1e-4

    # Adjust batch size per device (BS) and number of devices (ND) according to model size
    # Number of devices (ND) is determined by a combination of factors, context length, accelerator type, whether LoRA is used, etc.
    if size == "7b" or size == "13b":  
        # nd is set on the basis of using Nvidia A10 and conducting full parameter fine-tuning with default context length.
        # If Nvidia A100 is used. ND can be set to 8 instead.
        bs, nd = 16, 16
    elif size == "70b":
        # nd is set on the basis of using Nvidia A10 and conducting full parameter fine-tuning with default context length.
        # If Nvidia A100 is used. ND can be set to 16 instead.
        bs, nd = 8, 32
    else:
        print(f"Invalid size: {size}")
        sys.exit(1)

    # Pick model ID and deepspeed config based on model size
    model_id = f"meta-llama/Llama-2-{size}-hf"
    config_dir = f"./deepspeed_configs/zero_3_llama_2_{size}.json"

    # Update the config with the adjusted parameters
    config.update(
        **{
            "as_test": args.as_test,
            "batch_size": bs,
            "batch_size_per_device": args.batch_size_per_device,
            "block_size": args.ctx_len,
            "ds_config": config_dir,
            "eval_batch_size": args.eval_batch_size_per_device,
            "grad_accum": args.grad_accum,
            "lora": args.lora,
            "lr": lr,
            "mixed_precision": args.mixed_precision,
            "model_name": model_id,
            "no_grad_ckpt": args.no_grad_ckpt,
            "num_checkpoints_to_keep": args.num_checkpoints_to_keep,
            "num_devices": nd,
            "num_epochs": args.num_epochs,
            "output_dir": args.output_dir,
            "seed": 42,
            "special_token_path": args.special_token_path,
            "stop_perplexity": args.stop_perplexity,
            "train_path": args.train_path,
            "test_path": args.test_path,
        }
    )
```

Comment (ArturNiederfahrenhorst) [reply templates/fine-tune-llama2/train.py:684]: Sorry I didn't catch this earlier.

**Review by ArturNiederfahrenhorst — COMMENTED** [review]
Thanks for the changes!

I only have small nits. Can you make these small changes, and run 7B, 13B an 70B with LoRA one last time and post a screenshot here that "prooves" that everything works on one of these trainings? Then we can merge I'd say!

File: templates/fine-tune-llama2/README.md (hunk 8 above)
Comment (scottsun94) [reply templates/fine-tune-llama2/README.md]: Done

File: configs/fine-tune-llama2/gce.yaml (hunk 17 above)
Comment (scottsun94) [reply configs/fine-tune-llama2/gce.yaml:7]: 1. This compute config doesn't work for 70B for sure because you need g5.48xlarge for that. The 2 configs are only for 7B. In the README, we do ask people to switch to different instance types manually if they want to train for 13B and 70B.
2. For AWS, I can use g5.4xlarge for 7b, g5.12xlarge for 13b, g5.48xlarge for 70B. I'm not sure about GCE, the old template doesn't say anything about what instance types to use for 13b and 70b. Can you advice?



RE: 1, I've kicked off full parameter again. Let's see how it goes.
RE: 2. Let's not block on GCE yaml because AIOA will only work on AWS for now. Let's fix it after we merge it.

Comment (ArturNiederfahrenhorst) [reply configs/fine-tune-llama2/gce.yaml:7]: Sure thing! Please post a screenshot of the successful training from "RE: 1"

Comment (scottsun94) [reply configs/fine-tune-llama2/gce.yaml:7]: 6 runs were all successful: (7b, 13b, 70b) x (lora, without lora)
- 7b full parameter
Screenshot 2024-01-15 at 9 28 31 PM

-7b lora
Screenshot 2024-01-15 at 9 28 45 PM

-13b full parameter

Screenshot 2024-01-15 at 9 29 14 PM


-13b lora
Screenshot 2024-01-15 at 9 29 27 PM


-70b full parameter
Screenshot 2024-01-16 at 9 06 34 AM
 
Screenshot 2024-01-16 at 9 06 38 AM

-70b lora
Screenshot 2024-01-16 at 9 07 15 AM

Screenshot 2024-01-16 at 9 07 20 AM

Comment (scottsun94) [reply configs/fine-tune-llama2/gce.yaml:7]: BTW, I also did a 7b full parameter run using the gcp yaml. It worked as well. The only issue is that the l4 gpu seems much harder to get.
Screenshot 2024-01-16 at 9 13 33 AM

**Review by ArturNiederfahrenhorst — APPROVED** [review]

File: configs/fine-tune-llama2/gce.yaml (hunk 17 above)
Comment (ArturNiederfahrenhorst) [reply configs/fine-tune-llama2/gce.yaml:7]: Thanks for doing all this work. Great to see that everything works as expected 👍
